libraries/Microsoft.Bot.Connector/Authentication/BotFrameworkClientImpl.cs

```diff
@@ -0,0 +1,131 @@
+
+            var credentials = await _credentialsFactory.CreateCredentialsAsync(fromBotId, toBotId, _loginEndpoint, true, cancellationToken).ConfigureAwait(false);
+
+            // Clone the activity so we can modify it before sending without impacting the original object.
+            var activityClone = JsonConvert.DeserializeObject<Activity>(JsonConvert.SerializeObject(activity));
+
+            // Apply the appropriate addressing to the newly created Activity.
+            activityClone.RelatesTo = new ConversationReference
+            {
+                ServiceUrl = activityClone.ServiceUrl,
+                ActivityId = activityClone.Id,
+                ChannelId = activityClone.ChannelId,
+                Locale = activityClone.Locale,
+                Conversation = new ConversationAccount
+                {
+                    Id = activityClone.Conversation.Id,
+                    Name = activityClone.Conversation.Name,
+                    ConversationType = activityClone.Conversation.ConversationType,
+                    AadObjectId = activityClone.Conversation.AadObjectId,
+                    IsGroup = activityClone.Conversation.IsGroup,
+                    Properties = activityClone.Conversation.Properties,
+                    Role = activityClone.Conversation.Role,
+                    TenantId = activityClone.Conversation.TenantId,
+                }
+            };
+            activityClone.Conversation.Id = conversationId;
+            activityClone.ServiceUrl = serviceUrl.ToString();
+            activityClone.Recipient ??= new ChannelAccount();
+            activityClone.Recipient.Role = RoleTypes.Skill;
+
+            // Create the HTTP request from the cloned Activity and send it to the Skill.
+            using (var jsonContent = new StringContent(JsonConvert.SerializeObject(activityClone, new JsonSerializerSettings { NullValueHandling = NullValueHandling.Ignore }), Encoding.UTF8, "application/json"))
+            {
+                using (var httpRequestMessage = new HttpRequestMessage())
+                {
+                    httpRequestMessage.Method = HttpMethod.Post;
+                    httpRequestMessage.RequestUri = toUrl;
+                    httpRequestMessage.Content = jsonContent;
+
+                    // Add the auth header to the HTTP request.
+                    await credentials.ProcessHttpRequestAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
+
+                    using (var httpResponseMessage = await _httpClient.SendAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false))
+                    {
+                        var content = httpResponseMessage.Content != null ? await httpResponseMessage.Content.ReadAsStringAsync().ConfigureAwait(false) : null;
+
+                        // Capture the HTTP response, assuming JSON in the response body, otherwise make it NULL.
+                        return new InvokeResponse<T>
```

Comment:
I found some issues with this code. When things fail, `GetBodyContent` does not have `T` shape, instead it's general a string error message. A more correct approach might be to check httpResponseMessage.StatusCode, if it is not a success status code, get the error info from the body if any and throw an exception containing that error info. Do NOT try to deserialize the body into T when status code is non-success because it will always fail and the error message will get swollen. I ran into this once this week and it's almost impossible to debug since you never get notifiedof the error and just get 'default'.

Reply:
Sorry I just took this code from the existing BotFrameworkHttpClient verbatim - but what you say is absolutely correct - lets fix it now.